lib/IRGen/IRGenModule.cpp

```diff
@@ -1185,63 +1175,186 @@ static bool isFirstObjectFileInModule(IRGenModule &IGM) {
   return containingModule->getFiles().front() == file;
 }
 
+static bool
+doesTargetAutolinkUsingAutolinkExtract(const SwiftTargetInfo &TargetInfo,
+                                       const llvm::Triple &Triple) {
+  if (TargetInfo.OutputObjectFormat == llvm::Triple::ELF && !Triple.isPS4())
+    return true;
+
+  if (TargetInfo.OutputObjectFormat == llvm::Triple::Wasm)
+    return true;
+
+  if (Triple.isOSCygMing())
+    return true;
+
+  return false;
+}
+
+namespace {
+struct AutolinkKind {
+  enum ValueTy {
+    LLVMLinkerOptions,
+    LLVMDependentLibraries,
+    SwiftAutoLinkExtract,
+  };
+
+  ValueTy Value;
+
+  AutolinkKind(ValueTy value) : Value(value) {}
+  AutolinkKind(const AutolinkKind &kind) : Value(kind.Value) {}
+  operator enum ValueTy() { return Value; }
+
+  StringRef getSectionNameMetadata() {
+    // FIXME: This constant should be vended by LLVM somewhere.
+    switch (Value) {
+    case AutolinkKind::LLVMDependentLibraries:
+      return "llvm.dependent-libraries";
+      break;
+    case AutolinkKind::LLVMLinkerOptions:
+    case AutolinkKind::SwiftAutoLinkExtract:
+      return "llvm.linker.options";
+      break;
+    }
+  }
+
+  void collectEntriesFromLibraries(
+      llvm::SmallVector<llvm::MDNode *, 4> &Entries,
+      const SmallVector<LinkLibrary, 32> &AutolinkEntries, IRGenModule &IGM) {
+    llvm::LLVMContext &ctx = IGM.getLLVMContext();
+
+    switch (Value) {
+    case AutolinkKind::LLVMLinkerOptions:
+    case AutolinkKind::SwiftAutoLinkExtract: {
+      // On platforms that support autolinking, continue to use the metadata.
+      for (LinkLibrary linkLib : AutolinkEntries) {
+        switch (linkLib.getKind()) {
+        case LibraryKind::Library: {
+          llvm::SmallString<32> opt =
+              getTargetDependentLibraryOption(IGM.Triple, linkLib.getName());
+          Entries.push_back(
+              llvm::MDNode::get(ctx, llvm::MDString::get(ctx, opt)));
+          break;
```

Comment:
Rather than breaking here, my suggestion is to continue. And then at the bottom of the switch put in an llvm_unreachable like above.